geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java

```diff
@@ -237,15 +238,21 @@ public int hstrlen(byte[] field) {
     return new ArrayList<>(hash.keySet());
   }
 
-  public ImmutablePair<Integer, List<byte[]>> hscan(Pattern matchPattern,
-      int count,
-      int cursor) {
-
-    ArrayList<byte[]> resultList = new ArrayList<>(count + 2);
+  public ImmutablePair<Integer, List<byte[]>> hscan(Pattern matchPattern, int count, int cursor) {
+    // No need to allocate more space than it's possible to use given the size of the hash. We need
+    // to add 1 to hash.size() to ensure that if count > hash.size(), we return a cursor of 0
+    long maximumCapacity = 2L * Math.min(count, hlen() + 1);
+    if (maximumCapacity > Integer.MAX_VALUE) {
```

Comment:
Where you able to test this in the integration test? I know you did in the unit test but it is hard to do when you need  the hash to have more than 1G entries. If you could use the same hlen() trick in the integration test that you used in the unit test then you could test how this exception will be handled by the higher levels of the code.

Reply:
The integration tests are designed to simulate the behaviour of a client connecting to a Redis cluster, and to be agnostic about whether that cluster is a Geode Radish cluster or a native Redis cluster, so it's not really possible to get under the hood and mock out the returns of methods in our classes to force certain behaviours, unfortunately.